File: activemq-syscall/src/main/java/org/apache/activemq/syscall/util/DiskBenchmark.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.activemq.syscall.util;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.RandomAccessFile;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

import org.apache.activemq.syscall.FileDescriptor;
import org.apache.activemq.syscall.NativeAllocation;
import org.apache.activemq.util.cli.CommonsCLISupport;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import org.apache.commons.cli.PosixParser;

import static java.util.concurrent.TimeUnit.*;
import static org.apache.activemq.syscall.jni.IO.*;
import static org.apache.activemq.util.cli.OptionBuilder.*;

/**
 * This class is used to get a benchmark the raw disk performance.
 */
public class DiskBenchmark {

    private static final HashSet<String> OPS = new HashSet<String>();
    private static String op(String v) {
        OPS.add(v);
        return v;
    }
    private static final String WRITE_OP = op("write");
    private static final String READ_OP = op("read");
    
    private static final HashSet<String> APIS = new HashSet<String>();
    private static String api(String v) {
        APIS.add(v);
        return v;
    }
    private static final String NATIVE_BIO_API = api("native-bio");
    private static final String JAVA_API = api("java");
    
    boolean verbose;
    // reads and writes work with 4k of data at a time.
    int blockSize=1024*4; 
    long blocks=128000;
    String file;
    int sampleCount = 10;
    double samplePeriod = 1;
    
    String[] api = { JAVA_API, NATIVE_BIO_API };
    String[] operation = { WRITE_OP };
    boolean[] random={ false };
    boolean[] memcopy={ false };
    boolean[] sync={ false };
    
    static private Options createOptions() {
        Options options = new Options();
        options.addOption("h", "help",    false, "Display help information");
        options.addOption("v", "verbose", false, "Verbose");

        options.addOption(ob()
                .id("f")
                .name("file")
                .arg("path")
                .description("").op());

        options.addOption(ob()
                .id("s")
                .name("block-size")
                .arg("bytes")
                .description("").op());

        options.addOption(ob()
                .id("b")
                .name("blocks")
                .arg("count")
                .description("").op());

        options.addOption(ob()
                .id("c")
                .name("sample-count")
                .arg("number")
                .description("").op());
        
        options.addOption(ob()
                .id("p")
                .name("sample-period")
                .arg("seconds")
                .description("").op());

        options.addOption(ob()
                .name("sync")
                .args(2)
                .sperator(',')
                .arg("true,false")
                .description("").op());

        return options;
    }


    public static void main(String[] args) {
        String jv = System.getProperty("java.version").substring(0, 3);
        if (jv.compareTo("1.5") < 0) {
            System.err.println("This application requires jdk 1.5 or higher to run, the current java version is " + System.getProperty("java.version"));
            System.exit(-1);
            return;
        }

        DiskBenchmark app = new DiskBenchmark();
        System.exit(app.execute(args));
    }
    
    ///////////////////////////////////////////////////////////////////
    // Entry point for an embedded users who want to call us with
    // via command line arguments.
    ///////////////////////////////////////////////////////////////////
    public int execute(String[] args) {
        CommandLine cli = null;
        try {
            cli = new PosixParser().parse(createOptions(), args, true);
        } catch (ParseException e) {
            System.err.println( "Unable to parse command line options: " + e.getMessage() );
            displayHelp();
            return 1;
        }

        if( cli.hasOption("h") ) {
            displayHelp();
            return 0;
        }
        
        DiskBenchmark benchmark = new DiskBenchmark();
        args = CommonsCLISupport.setOptions(benchmark, cli);
        try {
            benchmark.benchmark();
            return 0;
        } catch (UsageException e) {
            System.out.println("Invalid Usage: "+e.getMessage());
            displayHelp();
            return -1;
        } catch (Throwable e) {
            System.out.flush();
            if (benchmark.verbose) {
                System.err.println("ERROR:");
                e.printStackTrace();
            } else {
                System.err.println("ERROR: " + e);
            }
            System.err.flush();
            return -2;
        }
    }
    
    @SuppressWarnings("serial")
    static class UsageException extends Exception {
        public UsageException(String msg) {
            super(msg);
        }
    }
    
    public void benchmark() throws IOException, UsageException {
        if( file==null ) {
            throw new UsageException("file not specified.");
        }
        
        for (String v : api) {
            if( !APIS.contains(v) ) {
                throw new UsageException("invalid api: "+v);
            }
        }
        for (String v : operation) {
            if( !OPS.contains(v) ) {
                throw new UsageException("invalid operation: "+v);
            }
        }
        
        PrintWriter pw = new PrintWriter(new FileOutputStream("report.html"));
        writeReportHeader(pw);
        int chartCounter=0;
        for (boolean memcopy : this.memcopy) {
            for (String operation : this.operation) {
                for (boolean random : this.random) {
                    for (boolean sync : this.sync) {
                        ArrayList<Benchmark> results = new ArrayList<Benchmark>(this.api.length); 
                        for (String apiName : this.api) {
                            Benchmark benchmark = createBenchmark(apiName);
                            benchmark.random = random;
                            benchmark.sync = sync;
                            benchmark.memcopy = memcopy;
                            benchmark.apiName = apiName;
                            benchmark.operation = operation;
                            benchmark.execute();
                            results.add(benchmark);
                        }
                        writeReportChart(pw, chartCounter++, results);
                    }
                }
            }
        }
        
        writeReportFooter(pw);
        pw.close();

    }

    private void writeReportHeader(PrintWriter pw) {
        if(pw==null) 
            return;
        pw.println("<html>");
        pw.println("  <head>");
        pw.println("    <script type='text/javascript' src='http://www.google.com/jsapi'></script>");
        pw.println("    <script type='text/javascript'>");
        pw.println("      google.load('visualization', '1', {'packages':['linechart']});");
        pw.println("    </script>");
        pw.println("    <style type='text/css'>");
        pw.println("      .chart-section {width:640px; padding-left: 30px}");
        pw.println("      .chart-props {width:140px; padding:0; padding-top:20px; float:left;}");
        pw.println("      .chart-graph {width: 500px; height: 200px; float:right; }");
        pw.println("    </style>");
        pw.println("  </head>");
        pw.println("  <body>");
    }
    
    private void writeReportChart(PrintWriter pw, int id, ArrayList<Benchmark> data) {
        if(pw==null) 
            return;
        Benchmark d1 = data.get(0);
        
        String titleX = String.format("Period (%.2f seconds)", samplePeriod);
        String titleY = "IO Operations";
        
        pw.println("    <div class='chart-section'>");
        pw.println("      <ul class='chart-props'>");
        pw.println("        <li>operation: "+d1.operation+"</li><li>sync: "+d1.sync+"</li><li>memcopy: "+d1.memcopy+"</li><li>random: "+d1.random+"</li>");
        pw.println("      </ul>");
        pw.println("      <div id='chart_"+id+"' class='chart-graph '></div>");
        pw.println("    </div>");
        pw.println("    <script type='text/javascript'>");
        pw.println("      google.setOnLoadCallback(draw_chart_"+id+");");
        pw.println("      function draw_chart_"+id+"() {");
        pw.println("        var data = new google.visualization.DataTable();");
        pw.println("        data.addColumn('string', 'Period');");
        
        for (Benchmark d : data) {
            pw.println("        data.addColumn('number', '"+d.apiName+"');");
        }
        pw.println("        data.addRows([");
        for( int i=0; i < sampleCount; i++ ) {
            if( i!=0 ) {
                pw.println(",");
            }
            pw.print("          ['"+i+"'");
            for (Benchmark d : data) {
                long value = 0;
                if( d.samples.size() >i ) {
                    value = d.samples.get(i);
                }
                pw.print(", "+value);
            }
            pw.print("]");
        }
        
        pw.println("        ]);");
        pw.println("        var chart = new google.visualization.LineChart(document.getElementById('chart_"+id+"'));");
        pw.println("        chart.draw(data, {legend: 'bottom', smoothLine:true, titleX:'"+titleX+"', titleY:'"+titleY+"' });");
        pw.println("      }");
        pw.println("    </script>");
        
    }

    private String title(ArrayList<Benchmark> data) {
        Benchmark d = data.get(0);
        return "operation: "+d.operation+", sync: "+d.sync+", memcopy: "+d.memcopy+", random: "+d.random;
    }


    private void writeReportFooter(PrintWriter pw) {
        if(pw==null) 
            return;
        pw.println("  </body>");
        pw.println("</html>");
    }

    private Benchmark createBenchmark(String api) {
        if( JAVA_API.equals(api) ) 
            return new JavaApi();
        if( NATIVE_BIO_API.equals(api) ) 
            return new NativeBioApi();
        throw new RuntimeException("Unsupported API: "+api);
    }

    static public class Sampler extends Thread {
        private final AtomicReference<ArrayList<Long>> samples = new AtomicReference<ArrayList<Long>>();
        private final AtomicLong metric;
        private final int count;
        private final long period;
        public boolean verbose;
        
        public Sampler(AtomicLong metric, int count, double periodInSecs) {
            super("Sampler");
            this.metric = metric;
            this.count = count;
            this.period = ns(periodInSecs);
            setPriority(MAX_PRIORITY);
            setDaemon(true);
        }
        
        static private long ns(double v) {
            return (long)(v*NANOSECONDS.convert(1, SECONDS));
        }

        @Override
        public void run() {
            ArrayList<Long> samples = new ArrayList<Long>(count);
            try {
                long sleepMS = period/1000000;
                int sleepNS = (int) (period%1000000);
                long currentValue;
                long lastValue = metric.get();
                for (int i = 0; i < count; i++) {
                    if( verbose ) {
                        System.out.print(".");
                    }
                    Thread.sleep(sleepMS,sleepNS);
                    currentValue = metric.get();
                    samples.add(currentValue-lastValue);
                    lastValue=currentValue;
                }
            } catch (InterruptedException e) {
            } finally {
                if( verbose ) {
                    System.out.println();
                }
                this.samples.set(samples);
            }
        }
        
        public synchronized ArrayList<Long> getSamples() {
            return samples.get();
        }
    }
    
    abstract class Benchmark {
        
        public String apiName;
        public String operation;
        public boolean random; 
        public boolean sync;
        public boolean memcopy;
        public ArrayList<Long> samples;
        
        final public void execute() throws IOException {
            
            boolean write;
            if( WRITE_OP.equals(operation) ) {
                write = true;
            } else if( READ_OP.equals(operation) ) {
                write = false;
            } else {
                throw new RuntimeException();
            }
            
            System.out.print("Benchmarking: api: "+apiName+", operation: "+operation+", random: "+random+", sync: "+sync+", memcopy: "+memcopy+" ");

            AtomicLong ioCount=new AtomicLong();
            Sampler sampler = new Sampler(ioCount, sampleCount, samplePeriod);
            sampler.verbose = verbose;
            try {
                init(true);
                sampler.start();
                outer: while( true ) {
                    seek(0);
                    for( long i=0; i < blocks; i++) {
                        if( write ) {
                            write();
                            if( sync ) {
                                sync();
                            }
                        } else {
                            read();
                        }
                        ioCount.incrementAndGet();
                        if( !sampler.isAlive() ) {
                            break outer;
                        }
                    }
                }
            } finally {
                dispose();
            }
            samples = sampler.getSamples();
            System.out.println("  results: "+samples);
            System.out.println();

        }

        abstract protected void init(boolean write) throws IOException;
        abstract protected void dispose() throws IOException;
        abstract protected void seek(long pos) throws IOException;
        abstract protected void write() throws IOException;
        abstract protected void read() throws IOException;
        abstract protected void sync() throws IOException;
        
    }
    
    class NativeBioApi extends Benchmark {
        
        private FileDescriptor fd;
        private NativeAllocation data;

        @Override
        protected void init(boolean write) throws IOException {
            data = NativeAllocation.allocate(block());
            if( write ) {
                int oflags =  O_CREAT | O_TRUNC | O_WRONLY;
                int mode = S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH;
                fd = FileDescriptor.open(file, oflags, mode);
            } else {
                int oflags =  O_RDONLY;
                fd = FileDescriptor.open(file, oflags);
            }
        }

        @Override
        protected void dispose() throws IOException {
            if( data!=null ) 
                data.free();
            if( fd!=null ) 
                fd.close();
        }

        @Override
        protected void seek(long pos) throws IOException {
            fd.seek(pos);
        }

        @Override
        protected void sync() throws IOException {
            fd.sync();
        }

        @Override
        protected void write() throws IOException {
            fd.write(data);
        }
        
        @Override
        protected void read() throws IOException {
            fd.read(data);
        }
    }
    
    class JavaApi extends Benchmark {
        
        private RandomAccessFile raf;
        private byte data[];

        @Override
        protected void init(boolean write) throws IOException {
            data = block();
            if( write ) {
                raf = new RandomAccessFile(file, "rw");
            } else {
                raf = new RandomAccessFile(file, "r");
            }
        }

        @Override
        protected void dispose() throws IOException {
            if( raf!=null ) 
                raf.close();
        }

        @Override
        protected void seek(long pos) throws IOException {
            raf.seek(pos);
        }

        @Override
        protected void sync() throws IOException {
            raf.getFD().sync();
        }

        @Override
        protected void write() throws IOException {
            raf.write(data);
        }
        
        @Override
        protected void read() throws IOException {
            raf.readFully(data);
        }
    }
    
    private byte[] block() {
        byte []data = new byte[blockSize];
        for (int i = 0; i < data.length; i++) {
            data[i] = (byte)('a'+(i%26));
        }
        return data;
    }

    private void displayHelp() {
        System.err.flush();
        String app = System.getProperty("diskbenchmark.application");
        if( app == null ) {
            try {
                URL location = getClass().getProtectionDomain().getCodeSource().getLocation();
                String[] split = location.toString().split("/");
                if( split[split.length-1].endsWith(".jar") ) {
                    app = split[split.length-1];
                }
            } catch (Throwable e) {
            }
            if( app == null ) {
                app = getClass().getSimpleName();
            }
        }

        // The commented out line is 80 chars long.  We have it here as a visual reference
//      p("                                                                                ");
        p();
        p("Usage: "+ app +" [options]");
        p();
        p("Description:");
        p();
        pw("  "+app+" is a disk benchmarker.", 2);
        p();

        p("Options:");
        p();
        PrintWriter out = new PrintWriter(System.out);
        HelpFormatter formatter = new HelpFormatter();
        formatter.printOptions(out, 78, createOptions(), 2, 2);
        out.flush();
        p();
    }
    

    private void p() {
        System.out.println();
    }
    private void p(String s) {
        System.out.println(s);
    }
    private void pw(String message, int indent) {
        PrintWriter out = new PrintWriter(System.out);
        HelpFormatter formatter = new HelpFormatter();
        formatter.printWrapped(out, 78, indent, message);
        out.flush();
    }


    public void setVerbose(boolean verbose) {
        this.verbose = verbose;
    }
    public void setBlockSize(int blockSize) {
        this.blockSize = blockSize;
    }
    public void setBlocks(long blocks) {
        this.blocks = blocks;
    }
    public void setFile(String file) {
        this.file = file;
    }
    public void setSampleCount(int sampleCount) {
        this.sampleCount = sampleCount;
    }
    public void setSamplePeriod(double samplePeriod) {
        this.samplePeriod = samplePeriod;
    }
    public void setApi(String[] api) {
        this.api = api;
    }
    public void setOperation(String[] operation) {
        this.operation = operation;
    }
    public void setRandom(boolean[] random) {
        this.random = random;
    }
    public void setMemcopy(boolean[] memcopy) {
        this.memcopy = memcopy;
    }
    public void setSync(boolean[] sync) {
        this.sync = sync;
    }
    
}

File: apollo-cli/src/main/scala/org/apache/activemq/apollo/cli/commands/DiskBenchmark.scala
package org.apache.activemq.apollo.cli.commands

/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import io.airlift.command.{Arguments, Command, Option}
import java.io.{PrintStream, InputStream, RandomAccessFile, File}
import java.util.concurrent.TimeUnit
import javax.management.ObjectName
import management.ManagementFactory
import org.apache.activemq.apollo.util.{MemoryPropertyEditor, IOHelper}
import MemoryPropertyEditor._


class Report {
  
  var block_size: Int = 0
  var async_writes: Int = 0
  var async_write_duration: Long = 0L
  var sync_writes: Int = 0
  var sync_write_duration: Long = 0L
  var reads: Int = 0
  var read_duration: Long = 0L

  def async_write_size_rate: Float = {
    var rc: Float = async_writes
    rc *= block_size
    rc /= (1024 * 1024)
    rc /= (async_write_duration / 1000.0f)
    return rc
  }

  def async_write_rate: Float = {
    var rc: Float = async_writes
    rc /= (async_write_duration / 1000.0f)
    return rc
  }

  def sync_write_size_rate: Float = {
    var rc: Float = sync_writes
    rc *= block_size
    rc /= (1024 * 1024)
    rc /= (sync_write_duration / 1000.0f)
    return rc
  }

  def sync_write_rate: Float = {
    var rc: Float = sync_writes
    rc /= (sync_write_duration / 1000.0f)
    return rc
  }

  def read_size_rate: Float = {
    var rc: Float = reads
    rc *= block_size
    rc /= (1024 * 1024)
    rc /= (read_duration / 1000.0f)
    return rc
  }

  def read_rate: Float = {
    var rc: Float = reads
    rc /= (read_duration / 1000.0f)
    return rc
  }
  
  override def toString: String = {
    return "Async writes: \n" + "  " + async_writes + " writes of size " + block_size + " written in " + (async_write_duration / 1000.0) + " seconds.\n" +
            "  " + async_write_rate + " writes/second.\n" +
            "  " + async_write_size_rate + " megs/second.\n" +
            "\n" +
            "Sync writes: \n" + "  " + sync_writes + " writes of size " + block_size + " written in " + (sync_write_duration / 1000.0) + " seconds.\n" +
            "  " + sync_write_rate + " writes/second.\n" +
            "  " + sync_write_size_rate + " megs/second.\n" +
            "\n" +
            "Reads: \n" + "  " + reads + " reads of size " + block_size + " read in " + (read_duration / 1000.0) + " seconds.\n" +
            "  " + read_rate + " reads/second.\n" +
            "  " + read_size_rate + " megs/second.\n" +
            "\n" + ""
  }

}

object DiskBenchmark {
  
  final val PHYSICAL_MEM_SIZE = format((try {
    val mbean_server = ManagementFactory.getPlatformMBeanServer()
    mbean_server.getAttribute(new ObjectName("java.lang:type=OperatingSystem"), "TotalPhysicalMemorySize") match {
      case x:java.lang.Long=> Some(x.longValue)
      case _ => None
    }
  } catch {
    case _ => None
  }).getOrElse(1024*1024*500L))

}


/**
 * The apollo encrypt command
 */
@Command(name = "disk-benchmark", description = "Benchmarks your disk's speed")
class DiskBenchmark extends BaseAction {
  import DiskBenchmark._
  
  
  @Option(name = Array("--verbose"), description = "Enable verbose output")
  var verbose: Boolean = false
  @Option(name = Array("--sample-interval"), description = "The number of milliseconds to spend mesuring perfomance.")
  var sampleInterval: Long = 30 * 1000

  @Option(name = Array("--block-size"), description = "The size of each IO operation.")
  var block_size_txt = "4k"
  def block_size = parse(block_size_txt).toInt

  @Option(name = Array("--file-size"), description = "The size of the data file to use, this should be big enough to flush the OS write cache.")
  var file_size_txt = PHYSICAL_MEM_SIZE
  def file_size = parse(file_size_txt)

  @Option(name = Array("--warm-up-size"), description = "The amount of data we should initial write before measuring performance samples (used to flush the OS write cache).")
  var warm_up_size_txt = format(parse("500M").min(parse(PHYSICAL_MEM_SIZE)/2))
  def warm_up_size = parse(warm_up_size_txt)

  @Arguments(description="The file that will be used to benchmark your disk (must NOT exist)")
  var file = new File("disk-benchmark.dat")

  def execute(in: InputStream, out: PrintStream, err: PrintStream): Int = {
    init_logging
    try {
      if (file.exists) {
        out.println("File " + file + " allready exists, will not benchmark.")
      } else {
        out.println("Benchmark using data file: " + file.getCanonicalPath)

        // Initialize the block /w data..
        var data = new Array[Byte](block_size)
        var i: Int = 0
        while (i < data.length) {
          data(i) = ('a' + (i % 26)).asInstanceOf[Byte]
          i += 1
        }

        val report = new Report
        report.block_size = block_size

        // Pre-allocate the file size..
        var raf = new RandomAccessFile(file, "rw")
        try {

          out.println("Pre-allocating data file of size: "+file_size_txt)
          raf.setLength(file_size)
          raf.seek(file_size-1)
          raf.writeByte(0);
          IOHelper.sync(raf.getFD)

          if( warm_up_size > 0 ) {
            val max = warm_up_size
            out.println("Warming up... writing async "+warm_up_size_txt+" so that async writes don't have that much of an advantage due to the OS write cache.")
            write(raf, data, (count)=>{
              count > max
            })
          }


          out.println("Benchmarking async writes")
          var start = System.nanoTime()
          var end = start
          report.async_writes = write(raf, data, (count)=>{
            end = System.nanoTime
            TimeUnit.NANOSECONDS.toMillis(end-start) >  sampleInterval
          })
          report.async_write_duration = TimeUnit.NANOSECONDS.toMillis(end-start)

          out.println("Syncing previous writes before measuring sync write performance.. (might take a while if your OS has a big write cache)")
          IOHelper.sync(raf.getFD)

          out.println("Benchmarking sync writes")
          start = System.nanoTime()
          end = start
          report.sync_writes = write(raf, data, (count)=>{
            IOHelper.sync(raf.getFD)
            end = System.nanoTime
            TimeUnit.NANOSECONDS.toMillis(end-start) >  sampleInterval
          })
          report.sync_write_duration = TimeUnit.NANOSECONDS.toMillis(end-start)

          if(!filled) {
            file_size_txt = ""+raf.getFilePointer
            out.println("File was not fully written, read benchmark will be operating against: "+(file_size/(1024 * 1024.0f))+" megs of data" )
            raf.seek(0);
          }
          out.println("Benchmarking reads")
          start = System.nanoTime()
          end = start
          report.reads = read(raf, data, (count)=>{
            end = System.nanoTime
            TimeUnit.NANOSECONDS.toMillis(end-start) >  sampleInterval
          })
          report.read_duration = TimeUnit.NANOSECONDS.toMillis(end-start)

        } finally {
          out.println("Closing.")
          raf.close
        }
        file.delete
        out.println(report)
      }
      0
    } catch {
      case x:Helper.Failure=>
        sys.error(x.getMessage)
        1
      case e: Throwable =>
        if (verbose) {
          err.println("ERROR:")
          e.printStackTrace(System.out)
        } else {
          err.println("ERROR: " + e)
        }
      1
    }
  }
  
  var filled = false

  private def write(raf: RandomAccessFile, data: Array[Byte], until: (Long)=>Boolean) = {
    var file_position = raf.getFilePointer
    var counter = 0
    while (!until(counter.toLong * data.length)) {
      if(  file_position + data.length >= file_size ) {
        filled = true
        file_position = 0;
        raf.seek(file_position)
      }
      raf.write(data)
      counter += 1;
      file_position += data.length
    }
    counter
  }

  private def read(raf: RandomAccessFile, data: Array[Byte], until: (Long)=>Boolean) = {
    var file_position = raf.getFilePointer
    var counter = 0
    while (!until(counter.toLong * data.length)) {
      if( file_position + data.length >= file_size ) {
        file_position = 0;
        raf.seek(file_position)
      }
      raf.readFully(data)
      counter += 1;
      file_position += data.length
    }
    counter
  }

}